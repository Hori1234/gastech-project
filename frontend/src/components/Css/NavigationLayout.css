#components-layout-demo-top-side-2 .logo {
  width: 120px;
  height: 100px;
  background: rgba(153, 32, 32, 0.2);
  margin: 16px 28px 16px 0;
  float: left;
}

.site-layout-background {
  background: rgb(255, 255, 255);
  
}
.dividers-style{
    background: rgb(255, 255, 255);
    align-items: center;
}