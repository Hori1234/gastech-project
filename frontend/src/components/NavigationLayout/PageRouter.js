import React, { Suspense, lazy } from "react";
import { Redirect, Route, Switch } from "react-router-dom";
import PrivateRoute from "../PrivateRoute";
import { Spin } from "antd";
const UploadButton = lazy(() => import("../UploadButton/UploadButton"));
const AccountManagementLayout = lazy(() =>
  import("../AccountManagerLayout/AccountManagementLayout")
);
const Logout = lazy(() => import("../Logout/Logout"));
const ManualPlanning = lazy(() => import("../ManualPlanning/ManualPlanning"));
const ViewPlanning = lazy(() =>
  import("../ViewPlanningPageLayout/ViewPlanning")
);
const DataVisualization = lazy(() =>
  import("../DataVisualization/DataVisualization")
);
const MonthlyDataAnalytics = lazy(() =>
  import("../MonthlyDataAnalytics/MonthlyDataAnalytics")
);
const Home = lazy(() => import("../Home/Home"));
const WoocomerceComponent = lazy(() =>
  import("../WooComerceLayout/WoocomerceComponent")
);
const WoocomerceOTMDcn = lazy(() =>
  import("../WooComerceLayout/WoocomerceOTMDcn")
);
const WoocomerceOTMDse = lazy(() =>
  import("../WooComerceLayout/WoocomerceOTMDse")
);

/**
 * Defines the routing for the pages in the NavigationLayout component.
 *
 * A private route only shows the component if the logged in user has the correct role.
 *
 * If a url is not matched with one of the routes, the user is redirected to the homepage.
 */
export default function PageRouter() {
  return (
    <Suspense
      fallback={
        <div
          style={{
            display: "flex",
            height: "100%",
            width: "100%",
            background: "white",
            justifyContent: "center",
            alignItems: "center",
          }}
        >
          <Spin size="large" />
        </div>
      }
    >
      <Switch>
        <PrivateRoute
          path="/upload"
          requiredRoles={["planner", "administrator"]}
        >
          <UploadButton />
        </PrivateRoute>
        <PrivateRoute path="/account" requiredRoles={["administrator"]}>
          <AccountManagementLayout />
        </PrivateRoute>
        <Route path="/logout">
          <Logout />
        </Route>
        <PrivateRoute
          path="/planning"
          requiredRoles={["planner", "administrator"]}
        >
          <ManualPlanning />
        </PrivateRoute>
        <PrivateRoute
          path="/view"
          requiredRoles={["view-only", "planner", "administrator"]}
        >
          <ViewPlanning />
        </PrivateRoute>
        <PrivateRoute path="/data" requiredRoles={["planner", "administrator"]}>
          <DataVisualization />
        </PrivateRoute>
        <PrivateRoute
          path="/monthly"
          requiredRoles={["view-only", "planner", "administrator"]}
        >
          <MonthlyDataAnalytics />
        </PrivateRoute>
        <PrivateRoute
          path="/"
          exact
          requiredRoles={["view-only", "planner", "administrator"]}
        >
          <Home />
        </PrivateRoute>
        <PrivateRoute
          path="/woocomcn"
          exact
          requiredRoles={["planner", "administrator"]}
        >
          <WoocomerceOTMDcn />
        </PrivateRoute>
        <PrivateRoute
          path="/woocomse"
          exact
          requiredRoles={["planner", "administrator"]}
        >
          <WoocomerceOTMDse />
        </PrivateRoute>
        <PrivateRoute
          path="/woocom"
          exact
          requiredRoles={["planner", "administrator"]}
        >
          <WoocomerceComponent />
        </PrivateRoute>
        <Route>
          <Redirect to="/" />
        </Route>
      </Switch>
    </Suspense>
  );
}
