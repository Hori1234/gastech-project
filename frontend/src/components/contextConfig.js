import React from "react";
import axios from "axios";
import { useHistory } from "react-router-dom";

// Context for keeping the current user state
export const AuthContext = React.createContext({
  status: "pending",
  error: null,
  user: null,
});

/**
 * Provides the Authentication context.
 *
 * Provides a component with the current user. It gives a component the ability to
 * login, logout, get the current user and verify the current users against a list of roles.
 *
 * @param props: The properties that need to set to the authentication provider
 */
const AuthProvider = (props) => {
  const [state, setState] = React.useState({
    status: "pending",
    error: null,
    user: null,
    getUsersStatus: "pending",
  });

  // Use history to redirect the user to the correct page after logging in
  let history = useHistory();

  /**
   * Get the current user from the backend.
   *
   * This effect only runs on the initial mount of this component.
   */
  React.useEffect((state) => {
    // check if user is already logged in
    axios
      .get("/auth/user")
      .then((res) => {
        setState((state) => ({
          ...state,
          user: res.data,
          status: "success",
        }));
      })
      .catch((error) => {
        setState((state) => ({
          ...state,
          status: "error",
          error: error,
        }));
      });
  }, []);

  /**
   * Try to log in a user using the credentials provided.
   *
   * @param credentials: An object containing the username, password and remember fields for the login.
   * @returns {Promise<boolean>}: Whether the login request was successful.
   */
  const login = async (credentials) => {
    return axios
      .post("/auth/login", credentials)
      .then((res) => {
        setState((state) => ({
          ...state,
          user: res.data,
          status: "success",
        }));
        return true;
      })
      .catch((error) => {
        setState((state) => ({
          ...state,
          status: "error",
          error: error,
        }));
        return false;
      });
  };

  /**
   * Try to log out the current user.
   *
   * @returns {Promise<boolean>}: Whether the logout request was successful.
   */
  const logout = async () => {
    return axios
      .post("/auth/logout")
      .then(() => {
        setState((state) => ({
          ...state,
          user: null,
          status: "logged-out",
        }));
        history.push("/");
        return true;
      })
      .catch((error) => {
        setState((state) => ({
          ...state,
          status: "error",
          error: error,
        }));
        return false;
      });
  };

  /**
   * Verify if the current user's role is in a list of roles.
   *
   * @param requiredRoles: A list of strings defining which roles are allowed.
   * @returns {boolean}: Whether the current user's role is in the requiredRoles list.
   */
  const verifyRole = (requiredRoles) => {
    if (state.user === null || state.user.role === null) {
      return false;
    } else {
      return requiredRoles.includes(state.user.role);
    }
  };

  // return the provider for the context, using the state and functions of this component
  return (
    <AuthContext.Provider
      value={{
        state,
        login,
        logout,
        verifyRole,
      }}
      {...props}
    />
  );
};

// Create a react hook for this context
const useAuth = () => React.useContext(AuthContext);

export { AuthProvider, useAuth };
