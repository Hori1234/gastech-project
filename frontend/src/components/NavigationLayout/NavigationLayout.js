import React, { useState } from "react";
import "antd/dist/antd.css";
import { Layout, Menu, Avatar, Divider, Image } from "antd";
import { UserOutlined } from "@ant-design/icons";
import { useLocation, Link } from "react-router-dom";
import "../Css/NavigationLayout.css";
import { useAuth } from "../contextConfig";
import PageRouter from "./PageRouter";
const { SubMenu } = Menu;
const { Content, Sider, Footer, Header } = Layout;

/**
 * Holds the sidebar and defines the main content layout.
 */
export default function NavigationLayout() {
  const auth = useAuth();
  const location = useLocation();
  const [state, setState] = useState({
    isEmptyState: true,
    UserState: false,
    collapsed: false,
    selectedKeys: [],
  });

  /**
   * Get the key of side menu item to highlight the current page.
   *
   * @param pathname: string of the current url path
   */
  const keys = (pathname) => {
    switch (pathname) {
      case "/":
        return "0";
      case "/account":
        return "1";
      case "/upload":
        return "2";
      case "/planning":
        return "3";
      case "/view":
        return "4";
      case "/data":
        return "5";
      case "/monthly":
        return "6";
      case "/woocom":
        return "8";
      case "/woocomsycc":
        return "9";
      default:
        return null;
    }
  };

  /** Make the sidebar collapsable
   *
   * @param collapsed: boolean to set the state of the sidebar
   */
  const onCollapse = (collapsed) => {
    setState({
      ...state,
      collapsed: collapsed,
    });
    const divider = document.getElementsByClassName("ant-divider")[0];

    if (collapsed) {
      divider.style.setProperty("display", "none");
    } else {
      divider.style.setProperty("display", "flex");
    }
  };

  return !auth.state ? (
    <pre>Loading...</pre>
  ) : (
    <Layout style={{ height: "100vh", fontFamily: "Trebuchet MS" }}>
      <Layout style={{ margin: 10 }}>
        <Sider
          collapsible
          collapsed={state.collapsed}
          onCollapse={onCollapse}
          width={200}
          className="site-layout-background"
        >
          <Layout className="dividers-style">
            <Avatar
              style={{ marginTop: 10 }}
              shape="square"
              size={64}
              icon={<UserOutlined />}
            />

            <Divider>O.T.M.D - Services</Divider>
            <Menu
              mode="inline"
              defaultOpenKeys={["sub1"]}
              style={{ height: "100%", borderRight: 0 }}
              selectedKeys={keys(location.pathname)}
            >
              {auth.state.user.role === "administrator" && (
                <SubMenu
                  key="sub1"
                  defaultSelectedKeys={["0"]}
                  icon={<UserOutlined />}
                  title={auth.state.user.username}
                >
                  <Menu.Item key="0">
                    <Link to="/">Home </Link>
                  </Menu.Item>
                  <Menu.Item key="1">
                    <Link to="/account">Account Management</Link>
                  </Menu.Item>
                  {/* <Menu.Item key="2">
                      <Link to="/upload">Upload</Link>
                    </Menu.Item>
                    <Menu.Item key="3">
                      <Link to="/planning">Create Planning</Link>
                    </Menu.Item>
                    <Menu.Item key="4">
                      <Link to="/view">View Planning</Link>
                    </Menu.Item>
                    <Menu.Item key="5">
                      <Link to="/data">Data Visualisation</Link>
                    </Menu.Item>
                    <Menu.Item key="6">
                      <Link to="/monthly">Monthly Data Analytics</Link>
                    </Menu.Item> */}
                  <Menu.Item key="8">
                    <Link to="/woocom">Woocomerce OTMD</Link>
                  </Menu.Item>
                  <Menu.Item key="9">
                    <Link to="/woocomsycc">Woocomerce Sycc</Link>
                  </Menu.Item>
                  <Menu.Item key="7" onClick={auth.logout}>
                    <Link to="/">Logout</Link>
                  </Menu.Item>
                </SubMenu>
              )}

              {auth.state.user.role === "planner" && (
                <SubMenu
                  key="sub1"
                  defaultSelectedKeys={["1"]}
                  icon={<UserOutlined />}
                  title={auth.state.user.username}
                >
                  <Menu.Item key="1">
                    <Link to="/">Home</Link>
                  </Menu.Item>
                  {/* <Menu.Item key="2">
                      <Link to="/upload">Upload</Link>
                    </Menu.Item>
                    <Menu.Item key="3">
                      <Link to="/planning">Create Planning</Link>
                    </Menu.Item>
                    <Menu.Item key="4">
                      <Link to="/view">View Planning</Link>
                    </Menu.Item>
                    <Menu.Item key="5">
                      <Link to="/data">Data Visualisation</Link>
                    </Menu.Item>
                    <Menu.Item key="6">
                      <Link to="/monthly">Monthly Data Analytics</Link>
                    </Menu.Item> */}
                  <Menu.Item key="7" onClick={auth.logout}>
                    <Link to="/">Logout</Link>
                  </Menu.Item>
                </SubMenu>
              )}

              {auth.state.user.role === "view-only" && (
                <SubMenu
                  key="sub1"
                  defaultSelectedKeys={["3"]}
                  icon={<UserOutlined />}
                  title={auth.state.user.username}
                >
                  <Menu.Item key="3">
                    <Link to="/">Home</Link>
                  </Menu.Item>
                  {/* <Menu.Item key="4">
                      <Link to="/view">View Planning</Link>
                    </Menu.Item>
                    <Menu.Item key="6">
                      <Link to="/monthly">Monthly Data Analytics</Link>
                    </Menu.Item> */}
                  <Menu.Item key="7" onClick={auth.logout}>
                    <Link to="/">Logout</Link>
                  </Menu.Item>
                </SubMenu>
              )}
            </Menu>
          </Layout>
        </Sider>
        <Layout style={{ padding: "0 24px 24px", height: "100%" }}>
          <Layout
            style={{
              marginTop: 5,
              minHeight: 280,
              alignItems: "center",
              width: "100%",
              height: "100%",
            }}
          >
            <Content
              className="site-layout-background"
              style={{
                padding: 24,
                marginTop: 5,
                minHeight: 280,
                width: "100%",
                height: "100%",
              }}
            >
              <Layout
                style={{
                  padding: 24,
                  width: "100%",
                  height: "100%",
                  alignItems: "center",
                  backgroundColor: "white",
                }}
              >
                <PageRouter />
              </Layout>
            </Content>
          </Layout>
          {/* <Footer style={{ textAlign: "center", height: 0 }}>
            O.T.M.D - Services Â©2021 Created by OTMD Group
          </Footer> */}
        </Layout>
      </Layout>
    </Layout>
  );
}
