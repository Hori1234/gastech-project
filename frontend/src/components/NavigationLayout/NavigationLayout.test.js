import React from 'react';
import { shallow } from 'enzyme';
import { AuthContext } from '../contextConfig';
import NavigationLayout from './NavigationLayout';


const ProvideUser = ({children}) => {
    const user = {
        state: {
            user: {
                id: 1,
                username: 'test',
                role: 'administrator'
            }
        }
    };
    return (
        <AuthContext.Provider value={user}>
            {children}
        </AuthContext.Provider>
    );
};


describe('NavigationLayout component', () => {


    it('shall shallow the login page correctly', () => {

        const component = shallow(
            <ProvideUser>
                <NavigationLayout />
            </ProvideUser>
        );
        expect(component).toMatchSnapshot();
    });

    /*
    it('shall have the correct view button href', () => {
        const component = shallow(
            <ProvideUser>
                <NavigationLayout/>
            </ProvideUser>
        );
        const viewButton = component.find('#viewButton > a')[0];
        expect(viewButton.attribs.href).toEqual('/view');

    });

    it('shall have the correct dataButton href', () => {
        const component = shallow(
            <ProvideUser>
                <NavigationLayout/>
            </ProvideUser>
        );
        const dataButton = component.find('#dataButton > a')[0];
        expect(dataButton.attribs.href).toEqual('/data');

    });

    it('shall have the correct monthlyButton href', () => {
        const component = shallow(
            <ProvideUser>
                <NavigationLayout/>
            </ProvideUser>
        );
        const monthlyButton = component.find('#monthlyButton > a')[0];
        expect(monthlyButton.attribs.href).toEqual('/montly');

    });

    it('shall have the correct uploadButton href', () => {
        const component = shallow(
            <ProvideUser>
                <NavigationLayout/>
            </ProvideUser>
        );
        const uploadButton = component.find('#uploadButton > a')[0];
        expect(uploadButton.attribs.href).toEqual('/upload');

    });

    it('shall have the correct planningButton href', () => {
        const component = shallow(
            <ProvideUser>
                <NavigationLayout/>
            </ProvideUser>
        );
        const planningButton = component.find('#planningButton > a')[0];
        expect(planningButton.attribs.href).toEqual('/planning');

    });

    it('shall have the correct accountButton href', () => {
        const component = shallow(
            <ProvideUser>
                <NavigationLayout/>
            </ProvideUser>
        );
        const accountButton = component.find('#accountButton > a')[0];
        expect(accountButton.attribs.href).toEqual('/account');
    });*/
});