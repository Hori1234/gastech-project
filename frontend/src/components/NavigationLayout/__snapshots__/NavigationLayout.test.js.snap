// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationLayout component shall shallow the login page correctly 1`] = `
<ContextProvider
  value={
    Object {
      "state": Object {
        "user": Object {
          "id": 1,
          "role": "administrator",
          "username": "test",
        },
      },
    }
  }
>
  <NavigationLayout />
</ContextProvider>
`;
